app-exp-dma.tex

```diff
@@ -0,0 +1,157 @@
+In-memory descriptors can often be chained or used in arrays in order to form \emph{scatter-gather lists} or \emph{ring buffers} to give a higher order structure to groups of transactions.
+
+\subsubsection{Bus-mastering DMA}
+In this case a device is able to make its own memory read and write transactions, independent of any system DMA controller.
+This is common for application-class devices such as PCI Express (PCIe) endpoints, including network cards, NVMe and AHCI storage, GPUs, NPUs and so on.
+In many cases a CHERI system cannot introspect into the device in order to modify its address behavior or to add CHERI support to its internal DMA infrastructure.
+Additionally, such devices may not be known at design-time any may change dynamically, thanks to hotpluggable PCIe cards and Thunderbolt devices.
+
+In order to safely constrain such a device, we must interpose something between the device and the memory subsystem.
+This is necessary in order to both limit the damage the device can do, and to protect the integrity of a CHERI system.
+Current systems use an IOMMU (also called a System MMU) to apply a page-based protection and translation model.
+This has significant performance implications due to the highly dynamic behavior of complex devices.
+It may be desirable to augment or replace the IOMMU with some other structure which instead uses the capability model, along the same lines of using CHERI for protection and the MMU for translation.
+Examples of such schemes are given in Cheng and Markettos~\cite{DBLP:conf/micro/MarkettosBBNMW20, cheng2025-capchecker}.
+
+Where possible, a preferable design point is to integrated CHERI support into a bus-mastering device such that the hardware can be trusted to manipulate capabilities correctly.
+In this instance we are able to run untrusted I/O software stack safely on trustworthy CHERI-enabled hardware.
+An example of this is a CHERI-enabled GPU~\cite{UCAM-CL-TR-997,naylor26-simtight}, which trustworthily shares capabilities with a CHERI-enabled CPU core.
+
+\subsection{Composing CHERI and DMA}
+DMA controllers operate outside the world of CPU-derived capabilities.
+The question arises how to compose them.
+Below we describe some different levels of utilizing capabilities with DMA.
+
+\subsubsection{Level 0: DMA clears tags}
+The simplest yet safe scenario is that DMA operates outside of the world of a CHERI-enabled CPU.
+The DMA hardware is unchanged, using integer address rather than capabilities.
+Any data written by the DMA clears tags, so the DMA is unable to forge or unsafely modify valid capabilities.
+Any pre-existing peripheral device can safely operate in this manner, although this also means they will not follow the capability protection model; an IOMMU may be required to safely constrain their memory access.
+
+\subsubsection{Level 1: addresses replaced by capabiities}
```

Comment:
FWIW, Nazerke Turtayeva did some work on this when she was an intern at MSR Cambridge a few years ago.  Her experiment with the CHERIoT Ibex core is open-source at https://github.com/microsoft/cheriot-ibex/tree/cheriot-dma .